### DISCUSSION TITLE
shaders for spine animation

### DISCUSSION THREAD

**Discussion by 10100010010001001010** [axmolengine/axmol#2233] [Q&A]
is it possible to apply a shader to spine animation? The examples in the tests looks confusing to me, for example there is a line like this:
`auto program = ProgramManager::getInstance()->loadProgram(positionTextureColor_vert, fragmentFilename, VertexLayoutType::Sprite);`

and I didn't find the SkeletonAnimation option in the VertexLayoutType. Should I get the layout manually via the spine extension?

**Comment by rh101** [comment]
What effect do you want to apply to the spine animation? Also, which examples in the tests are you referring to?

**Reply by 10100010010001001010** [reply]
dynamic lighting effect. As for the examples - the ShaderTest and ShaderTest2 files from axmol\tests\cpp-tests

**Reply by rh101** [reply]
Which shader, specifically? The reason is that, depending on the type of effect, it may or may not be possible.  

If you want to do something like what is described [here](https://en.esotericsoftware.com/forum/d/3770-dynamic-lighting-with-sprite-dlight), then that would be difficult, and it has nothing to do with Axmol, but rather because of the way Spine animation is constructed.  If you read the info at that link, then you'll understand why.

For anything else, you would just need to experiment to see if you can get the right effect that works on the spine animation.

**Comment by smilediver** [comment]
I think it is possible. You probably just need to set a custom shader through `SkeletonAnimation::setProgramState()`. `VertexLayoutType::Sprite` just means `V3F_C4B_T2F` vertex format and Spine uses the same format, so you can use that enum value when creating your custom shader.

Just keep in mind that two color tint Spines use `V3F_C4B_C4B_T2F` format, and I'm not sure how that would work with an override.
